[Static Analyzer CI] Reported results do not reflect changes to expectations in PRs

briannafan opened the pull request:
#### b3d26961a782dcdcc7fa912ffa4de9c4943e944f
<pre>
[Static Analyzer CI] Reported results do not reflect changes to expectations in PRs
<a href="https://rdar.apple.com/140374383">rdar://140374383</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284085">https://bugs.webkit.org/show_bug.cgi?id=284085</a>

Reviewed by Aakash Jain.

When expectations are changed in the PR, EWS results should abide by those expectations.
If not, we see inaccurate comments on PRs and miss potential regressions.

This change introduces logic that checks for modified expectations and will filter the
given results to conform to the expectations.
    1. If a file is removed from expectations, it will be treated as an unexpected failure.
       - If it passes, we suppress the pass as expected.
       - If it fails, we surface the failure, even if it was pre-existing.
    2. If a file is added to expectations, it will be treated as an expected failure.
       - If it passes, we surface the unexpected pass.
       - If it fails, we suppress the failure as expected.

* Tools/CISupport/ews-build/factories.py: Add FindModifiedSaferCPPExpectations to the factory.
(SaferCPPStaticAnalyzerFactory.__init__):
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/steps.py:
(FindModifiedSaferCPPExpectations):
    - New step that parses the diff of the PR to find which Safer CPP expectations
      were modified by the PR.
    - Tests are saved as &apos;user_removed_tests&apos; and &apos;user_added_tests&apos; build properties.
(FindModifiedSaferCPPExpectations.__init__):
(FindModifiedSaferCPPExpectations.run):
(FindModifiedSaferCPPExpectations.getResultSummary):

(FindUnexpectedStaticAnalyzerResults):
(FindUnexpectedStaticAnalyzerResults.run):
    - On this first run, we compare against the expectations in the PR. If the unexpected results
      match with the user modified tests, we can return results and skip a second build/results_db.
    - Otherwise, we check results_db but we ignore filtering tests if they are modified by the user.
(FindUnexpectedStaticAnalyzerResults.decode_results_data):
    - Load the log text in this function.
(FindUnexpectedStaticAnalyzerResults.get_unexpected_tests):
    - Simple function to convert a dictionary of results into a list of test strings.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db):
(FindUnexpectedStaticAnalyzerResults.check_results_db):
    - This is where we check for user modified tests. The logic is as follows:
      If a test is a pre-existing failure or pass after checking resultsdb, we
          1. Keep it as unexpected failure if a user removed the test
          2. Keep it as unexpected pass if a user added the test
          3. Remove it from unexpected if a user did not modify the test (normal behaviour)

(FindUnexpectedStaticAnalyzerResultsWithoutChange):
    - On the second run, we compare the build with PR changes to a build on ToT.
      If there are user modified tests, we filter the unexpected results and regenerate the results
      index if needed.
(FindUnexpectedStaticAnalyzerResultsWithoutChange.run):
(FindUnexpectedStaticAnalyzerResultsWithoutChange.filter_results_by_user_modification):
    - The filtering logic lives here!
    - For tests that the user added to expectations:
           1. If unexpected pass in the 1st run, add to unexpected passes here.
           2. If unexpected fail in the 2nd run, remove from unexpected failures here.
    - For tests that the user removed from expectations:
           1. If unexpected fail in the 1st run, add to unexpected failures here.
           2. If unexpected pass in the 2nd run, remove from unexpected passes here.
    - This filter ensures that the test results reflect the user modified expectations and not
      just which results were new to this PR.

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/290621@main">https://commits.webkit.org/290621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93d5711284386d60f28db21a090d850b6306c8ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90485 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/10017 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45421 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95496 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/41267 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92537 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/10410 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18336 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/95496 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/41267 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93486 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/10410 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/82068 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/95496 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/10410 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36420 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/40397 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/10410 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/37507 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97325 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/17677 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12982 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/97325 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/90070 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/17935 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77891 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97325 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20918 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10999 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14262 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/17687 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/17426 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/20881 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/19210 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7213,31 +7282,38 @@ def write_unexpected_results_file_to_master(self):
             f.write(results_data_obj)
 
-    def decode_results_data(self, string):
+    @defer.inlineCallbacks
```
Why is this async? I don't see anything we're calling that forces it to be.

briannafan replied:
I'm pretty sure that calling `yield self._addToLog` requires the function to be async. I remember I fixed this because I ran into an exception but I can't find the exact build.

JonWBedard commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -9249,6 +9249,69 @@ def test_success(self):
+        self.setProperty('buildnumber', 1234)
+
+        commit_diff = '''
+diff --git a/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations b/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations
```
Can we test a diff where a line is added and removed in the same file? Ideally one where the diff looks like a line was replaced. Something where we had a line that did this:
`inspector/agents/worker/WorkerAuditAgent.h -> inspector/agents/worker/WorkerWorkerAgent.h`

briannafan replied:
Yep, I will add a test for the following diff. Is that what you're looking for?
```
diff --git a/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations b/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations
index 8a2d2375b8d2..f7ebc3b11b94 100644
--- a/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations
+++ b/Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations
@@ -51,7 +51,7 @@ inspector/InspectorShaderProgram.h
 inspector/InspectorStyleSheet.cpp
 inspector/InspectorStyleSheet.h
 inspector/InspectorWebAgentBase.h
-inspector/agents/worker/WorkerAuditAgent.h
+inspector/agents/worker/WorkerWorkerAgent.h
```

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7358,10 +7445,8 @@ def run(self):
-        self.command += ['--delete-results']
+        self.command += ['--build-output', SCAN_BUILD_OUTPUT_DIR, '--archived-dir', os.path.join(self.getProperty('builddir'), 'build/baseline')]
+        self.command += ['--scan-build-path', '../llvm-project/clang/tools/scan-build/bin/scan-build', '--delete-results']  # Only generate results page on the second comparison
         if CURRENT_HOSTNAME in EWS_BUILD_HOSTNAMES + TESTING_ENVIRONMENT_HOSTNAMES and self.getProperty('github.base.ref', DEFAULT_BRANCH) == DEFAULT_BRANCH:
```
What are we changing here and why?

briannafan replied:
I just moved `'--delete-results'` one line up to be added with the other arguments to be succinct.
And moved the other arguments together too.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7138,7 +7138,61 @@ class ParseStaticAnalyzerResultsWithoutChange(ParseStaticAnalyzerResults):
+        if rc != SUCCESS:
+            return defer.returnValue(rc)
+
+        logText = self.log_observer.getStdout()
```
should you have: `wantStderr=True`

briannafan replied:
I don't think so, having the stdout seems sufficient?

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7138,7 +7138,61 @@ class ParseStaticAnalyzerResultsWithoutChange(ParseStaticAnalyzerResults):
+    command = ['git', 'diff', 'head~1', '--', '*Expectations']
+
+    def __init__(self, **kwargs):
+        super().__init__(logEnviron=False, **kwargs)
```
Do you want to add a timeout?

briannafan replied:
I think there is already a default timeout of 1200 seconds

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7138,7 +7138,61 @@ class ParseStaticAnalyzerResultsWithoutChange(ParseStaticAnalyzerResults):
+        self.setProperty('user_added_tests', added_tests)
+        return defer.returnValue(rc)
+
+    def getResultSummary(self):
```
what about the case when the step result is : FAILURE, EXCEPTION, CANCELLED etc.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7178,25 +7232,40 @@ def run(self):
+
+            results_json = yield self.decode_results_data()
+            if not results_json:
+                return defer.returnValue(False)
```
should it return or FAILURE (buildbot result code)?
Also, should add a log here to make it easy to debug if needed.

briannafan replied:
Good catch, it should return FAILURE.
There's a bit more logging in `decode_results_data`, I will add one more line there too.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7178,25 +7232,40 @@ def run(self):
+            unexpected_passes = self.get_unexpected_tests(results_json, 'passes') or ['Empty']
+            unexpected_failures = self.get_unexpected_tests(results_json, 'failures') or ['Empty']
+
+            if user_added_tests == unexpected_passes and user_removed_tests == unexpected_failures:
```
what if these lists have multiple elements? would it be the right way to compare the list?

briannafan replied:
No! I will change them to sets so the order doesn't matter.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7178,25 +7232,40 @@ def run(self):
+
+            if user_added_tests == unexpected_passes and user_removed_tests == unexpected_failures:
+                yield self._addToLog('stdio', 'Skipping second build since all unexpected results come from user modified expectations.\n')
+                unexpected_results_after_filter = True
```
should `unexpected_results_after_filter` be True or False here? Also, can we do early return from this if statement?

briannafan replied:
This should be True, since there are unexpected results (from line 7234, `if num_unexpected_results`), it just happens that they come from the user modified expectations and not from the checked in expectations.

I don't think we can exit early since there are various steps we want to append when there are unexpected results (around line 7269).

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7178,25 +7232,40 @@ def run(self):
+                unexpected_results_after_filter = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
+                if successful_filter and self.was_filtered and unexpected_results_after_filter:
+                    yield self._addToLog('stdio', f'\nSuccessfully filtered results! Updating unexpected_results.json on disk.\n')
+                    self.write_unexpected_results_file_to_master()
```
Is `write_unexpected_results_file_to_master` really creating file on the master, or creating locally on the bot. Please verify.

briannafan replied:
Yep it creates a file on the master. This function is also pre-existing code that got tested in a previous PR :)
```
bash-5.1$ ls public_html/results/macOS-Safer-CPP-Checks-EWS/cf77edef-15730/scan-build-output/
StaticAnalyzerUnexpectedRegressions  results.html  unexpected-results.html  unexpected_results.json
```

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7294,7 +7370,17 @@ def check_results_db(self, results_json, result_type, configuration, identifier)
+                    skip_filter = False
+                    if test_name in user_removed_tests and result_type == 'failures':
+                        skip_filter = True
+                    elif test_name in user_added_tests and result_type == 'passes':
```
should it be `elif`, or a single `if` with a `or` ?

briannafan replied:
A single `or` statement would suffice

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7388,15 +7472,65 @@ def run(self):
+
+        if user_removed_tests or user_added_tests:
+            unexpected_results_data = yield self.decode_results_data()
+            self.unexpected_results_filtered = unexpected_results_data
```
can merge this line with above, can skip creating an extra local variable.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7388,15 +7472,65 @@ def run(self):
+
+        for test_name in user_removed_tests:
+            split_test = test_name.split('/')
+            project, file, checker = split_test[0], '/'.join(split_test[1:-1]), split_test[-1]
```
should make sure that split_test[0] and split_test[1] etc. won't result in exception.

aj062 reviewed (approved):
rs=me

briannafan committed:
[Static Analyzer CI] Reported results do not reflect changes to expectations in PRs
rdar://140374383
https://bugs.webkit.org/show_bug.cgi?id=284085

Reviewed by Aakash Jain.

When expectations are changed in the PR, EWS results should abide by those expectations.
If not, we see inaccurate comments on PRs and miss potential regressions.

This change introduces logic that checks for modified expectations and will filter the
given results to conform to the expectations.
    1. If a file is removed from expectations, it will be treated as an unexpected failure.
       - If it passes, we suppress the pass as expected.
       - If it fails, we surface the failure, even if it was pre-existing.
    2. If a file is added to expectations, it will be treated as an expected failure.
       - If it passes, we surface the unexpected pass.
       - If it fails, we suppress the failure as expected.

* Tools/CISupport/ews-build/factories.py: Add FindModifiedSaferCPPExpectations to the factory.
(SaferCPPStaticAnalyzerFactory.__init__):
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):

* Tools/CISupport/ews-build/steps.py:
(FindModifiedSaferCPPExpectations):
    - New step that parses the diff of the PR to find which Safer CPP expectations
      were modified by the PR.
    - Tests are saved as 'user_removed_tests' and 'user_added_tests' build properties.
(FindModifiedSaferCPPExpectations.__init__):
(FindModifiedSaferCPPExpectations.run):
(FindModifiedSaferCPPExpectations.getResultSummary):

(FindUnexpectedStaticAnalyzerResults):
(FindUnexpectedStaticAnalyzerResults.run):
    - On this first run, we compare against the expectations in the PR. If the unexpected results
      match with the user modified tests, we can return results and skip a second build/results_db.
    - Otherwise, we check results_db but we ignore filtering tests if they are modified by the user.
(FindUnexpectedStaticAnalyzerResults.decode_results_data):
    - Load the log text in this function.
(FindUnexpectedStaticAnalyzerResults.get_unexpected_tests):
    - Simple function to convert a dictionary of results into a list of test strings.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db):
(FindUnexpectedStaticAnalyzerResults.check_results_db):
    - This is where we check for user modified tests. The logic is as follows:
      If a test is a pre-existing failure or pass after checking resultsdb, we
          1. Keep it as unexpected failure if a user removed the test
          2. Keep it as unexpected pass if a user added the test
          3. Remove it from unexpected if a user did not modify the test (normal behaviour)

(FindUnexpectedStaticAnalyzerResultsWithoutChange):
    - On the second run, we compare the build with PR changes to a build on ToT.
      If there are user modified tests, we filter the unexpected results and regenerate the results
      index if needed.
(FindUnexpectedStaticAnalyzerResultsWithoutChange.run):
(FindUnexpectedStaticAnalyzerResultsWithoutChange.filter_results_by_user_modification):
    - The filtering logic lives here!
    - For tests that the user added to expectations:
           1. If unexpected pass in the 1st run, add to unexpected passes here.
           2. If unexpected fail in the 2nd run, remove from unexpected failures here.
    - For tests that the user removed from expectations:
           1. If unexpected fail in the 1st run, add to unexpected failures here.
           2. If unexpected pass in the 2nd run, remove from unexpected passes here.
    - This filter ensures that the test results reflect the user modified expectations and not
      just which results were new to this PR.

* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/290621@main
